preview/FS-1068-open-type-declaration.md

```diff
@@ -0,0 +1,273 @@
+
+Additionally, important C# DSL APIs are starting to appear that effectively require this.
+
+## Detailed design
+[design]: #detailed-design
+
+Opening of types allows for treating a type somewhat as if it were a module with module-bound functions inside of it. For example:
+
+```fsharp
+type A() =
+    static member M(x: int) = x * 2
+
+open type A
+
+M(2)
+```
+
+However, **members are not functions**, so very different rules apply beyond simple cases like the above. This primarily concerns name resolution when overloads are involved, or when considering whether or not we resolve a method or a property first.
+
+Additionally, visible static fields should also be accessible. Example:
+
+```fsharp
+open type System.Data.Common.DbMetaDataColumnNames
+
+printfn "%s" ColumnSize
+```
+
+Visible consts should also be allowed. Example:
+
+```fsharp
+open type System.Math
+
+PI
+```
+
+### Attributes
+
+There are two relevant attributes that are also respected: `AutoOpen` and `RequireQualifiedAccess`.
+
+#### `RequireQualifiedAccess`
+
+As with modules, application of this attribute to a type in F# will require full qualification to use a member defined within.
+
+```fsharp
+[<RequireQualifiedAccess>]
+type A() =
+    static member M(x: int) = x * 2
+
+open type C // Compile error
+
+M(2) // 'M' is not recognized
+```
+
+#### `AutoOpen`
+
+Also like modules, specifying `AutoOpen` on a type automatically brings its members into scope.
+
+```fsharp
+[<AutoOpen>]
+type A() =
+    static member M(x: int) = x * 2
+
+M(2) // Can call 'M' without opening 'A'
+```
+
+### Opening types reveals nested types
+
+When opening a .NET type (or a provided type), any accessible nested types also become accessible without qualification.
+
+### Extending types
+
+It is possible to use optional type extensions to extend types. Members defined as type extensions are visible when opening these types:
+
+```fsharp
+type System.Math with
+    static member MyPI = 3.14
+
+open type System.Math
+
+MyPI // works!
+```
+
+Resolving members on optional type extensions works the same as with non-static extensions. These rules are not adjusted. That is, members defined in type extensions don't get magically attached to an already-opened static class. The following code will not compile because the `N` member is not in scope:
+
+```fsharp
+type A() =
+    static member M(x: int) = x * 2
+
+open type A
+
+type A with
+    static member N(x: float) = ()
+
+M(2)
+N(2.0) // Compile error, not in scope
+```
+
+The `open type A` declaration must be placed below the type extension:
+
+```fsharp
+type A() =
+    static member M(x: int) = x * 2
+
+type A with
+    static member N(x: float) = ()
+
+open type C
+
+M(2)
+N(2.0) // Works!
+```
+
+### Opening types with generic parameter instantiations
+
+You can open types with generic parameter instantiations:
```

Comment:
Can you open it multiple times with different instantiations? e.g.

```fsharp
open type Vector<int>
open type Vector<float>
```

Reply:
You can, but the latest one would shadow all nested types/method groups from the first one.